backend/lib/src/api/handlers.rs

```diff
@@ -169,19 +171,52 @@ pub async fn download_by_location(
     Ok(file_stream_resp.into_response())
 }
 
-pub async fn download_by_key(
+pub async fn internal_upload_by_key(
     State(_services): State<Services>,
-    TypedHeader(auth): TypedHeader<Authorization<Bearer>>,
-    Path((_bucket_id, _file_key)): Path<(String, String)>,
+    Path(file_key): Path<String>,
+    body: Bytes,
+) -> (StatusCode, impl IntoResponse) {
+    // let _auth = extract_bearer_token(&auth)?;
+    if let Err(e) = tokio::fs::create_dir_all("uploads").await {
+        return (StatusCode::INTERNAL_SERVER_ERROR, e.to_string());
+    }
+    match tokio::fs::write(format!("uploads/{}", file_key), body).await {
+        Ok(_) => (StatusCode::OK, "Upload successful".to_string()),
+        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()),
+    }
+}
+
+pub async fn download_by_key(
+    State(services): State<Services>,
+    Path(file_key): Path<String>,
 ) -> Result<impl IntoResponse, Error> {
-    let _auth = extract_bearer_token(&auth)?;
+    // TODO: re-add auth
+    // let _auth = extract_bearer_token(&auth)?;
 
-    // TODO(MOCK): return proper data
-    let file_data = b"Mock file content for download".to_vec();
-    let stream = ReaderStream::new(Cursor::new(file_data));
-    let file_stream_resp = FileStream::new(stream).file_name("by_key.txt");
+    let download_result = services.msp.get_file_from_key(&file_key).await?;
 
-    Ok(file_stream_resp.into_response())
+    // Extract filename from location or use file_key as fallback
+    let filename = download_result
+        .location
+        .split('/')
+        .last()
+        .unwrap_or(&file_key)
+        .to_string();
+
+    // Open file for streaming
+    let file = File::open(&download_result.temp_path)
+        .await
+        .map_err(|e| Error::BadRequest(format!("Failed to open downloaded file: {}", e)))?;
+
+    let stream = ReaderStream::new(file);
+    let file_stream_resp = FileStream::new(stream).file_name(&filename).into_response();
+
+    // Schedule cleanup of temp file after response
```

Comment:
This isn't scheduled tho - it will most likely just be done right after we return from this task.
The FileStream probably doesn't buffer the entire file in memory.